### DISCUSSION TITLE
Re: basic mp3 and/or sprite playback

### DISCUSSION THREAD

**Discussion by drscottlobo** [danigb/smplr#60] [Ideas]
Hey Danigb, still using smplr a ton in my music teaching app! It's working great with soundfonts. It would be awesome if I could also use smplr to playback short mp3 files or sprites. That way I wouldn't have to use a separate library when the audio is just an mp3 file, currently I'm using use-sound but it's pretty buggy and not maintained well. Any info is appreciated! 

Thanks.

**Comment by danigb** [comment]
Hi, 

Sorry for the late reply, and glad to hear that this is useful for you ❤️ . Audio sprites are not implemented yet, but open to do it. Can you describe precisely what you need or how do you use use-sound to achieve your goals?

**Reply by drscottlobo** [reply]
Hi danigb! I would not call that late at all! 

OK, so currently, I have a student give a note name to a pitch they see on the staff. If they get it correct, I use smplr to play a piano or guitar note of that pitch, if they get it wrong, I have a short mp3 that plays the "wrong" sound. Also at the end of completing an exercise, a short tune is played using a sprite depending on how well they did, a shorter tune for 1 star, a medium for 2 stars, the complete tune for 3 stars. I was using use-sound with a sprite that contained the wrong and ending tunes/ See mockup below: 

`
const sprite = './assets/stars.mp3'
const soundData = {
	"src": ["stars.ogg", "stars.m4a", "stars.mp3", "stars.ac3"],
	"sprite": {
		"one": [0, 2533.877551020408],
		"three": [4000, 4075.102040816326],
		"two": [10000, 3056.3265306122444],
                 "wrong": [15000, 2039.249249232533]
	}
}

const [play] = useSound(soundData, {volume});

function onEndExercise(score) {
if (score < 79) {
  play({id: 'one')
 } else if (score > 79 && score < 89) {
  play({id: 'two'})
 }  else { 
  play({id: 'three'})
 }
}
` 

Would be a cool thing if smplr could just handle the playback of those regions of an mp3 of sprites in a similar fashion. Maybe it already does if I can just load a custom mp3 file and then just define start point and duration?

Thanks so much for all your work on this!

**Reply by danigb** [reply]
Thanks for the detailed response. One thing I don't understand is why you need all this complexity. Why you can't use just a single file for each "tune" instead of pack them all inside a single file? I mean, I agree that could be a transport (network) optimization. But does it worth it? Maybe I'm missing something...

Said that, I'm open to implement sprites (won't be very soon). But I'm wondering if, at least for this case, it worth the effort...

**Comment by drscottlobo** [comment]
Wouldn’t really matter if it’s a sprite or just mp3 playback to me. I mean, there is certainly the extra network traffic of downloading multiple files vs a single file, no? Is it already possible to playback a single custom mp3 with smplr?My main goal is to have smplr be my only audio library in my music teaching app, and right now use sound is still needed for mp3 playback.Thanks!Scott _______________Scott Wolf, DMAOxnard College - Professor of MusicOn Feb 15, 2024, at 1:03 PM, danigb ***@***.***> wrote:﻿
Thanks for the detailed response. One thing I don't understand is why you need all this complexity. Why you can't use just a single file for each "tune" instead of pack them all inside a single file? I mean, I agree that could be a transport (network) optimization. But does it worth it? Maybe I'm missing something...
Said that, I'm open to implement sprites (won't be very soon). But I'm wondering if, at least for this case, it worth the effort...

**Comment by danigb** [comment]
Hey!

Yes, you can use a `Sampler` to load and play any mp3: https://github.com/danigb/smplr?tab=readme-ov-file#sampler

**Comment by drscottlobo** [comment]
It works great to play an mp3 with the Sampler! One issue, the onEnded()
callback doesn't work at the end of playing the mp3, perhaps because there
is no duration specified? Could you give access to the duration of the
loaded buffers? Or is there a way to make onEnded function when the mp3 has
completed playing?

Thanks!
Scott


On Sun, Feb 18, 2024 at 3:37 AM danigb ***@***.***> wrote:

> Hey!
>
> Yes, you can use a Sampler to load and play any mp3:
> https://github.com/danigb/smplr?tab=readme-ov-file#sampler
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/danigb/smplr/discussions/60#discussioncomment-8506675>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJVT3RG2YEA6OQL3O5NX7SLYUHRWFAVCNFSM6AAAAABC6WWOBSVHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM4DKMBWGY3TK>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
*Scott Wolf, DMA*
Oxnard College - Professor of Music
www.scottwolfguitar.com

All Strings Considered <http://www.scottwolfguitar.com/podcast.html>
Follow on Twitter - @allstrings <https://twitter.com/allstrings>
Subscribe free on iTunes
<https://itunes.apple.com/us/podcast/all-strings-considered/id562425515> -
Subscribe - Please Rate!

**Comment by drscottlobo** [comment]
Update: I can get access to the duration in a slightly hacky way:

const duration = sampler.player.player.player.buffers['id'].duration;

But then giving duration to the sampler note doesn't work either, as it
will no longer stop for some reason? sampler.start({note: 'id', duration})
won't stop playing with sampler.stop(), I'll submit a new issue with just a
request for onEnded to work after playing back of buffered mp3.

Scott


On Sun, Feb 18, 2024 at 10:00 AM Scott Wolf ***@***.***> wrote:

> It works great to play an mp3 with the Sampler! One issue, the onEnded()
> callback doesn't work at the end of playing the mp3, perhaps because there
> is no duration specified? Could you give access to the duration of the
> loaded buffers? Or is there a way to make onEnded function when the mp3 has
> completed playing?
>
> Thanks!
> Scott
>
>
> On Sun, Feb 18, 2024 at 3:37 AM danigb ***@***.***> wrote:
>
>> Hey!
>>
>> Yes, you can use a Sampler to load and play any mp3:
>> https://github.com/danigb/smplr?tab=readme-ov-file#sampler
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/danigb/smplr/discussions/60#discussioncomment-8506675>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AJVT3RG2YEA6OQL3O5NX7SLYUHRWFAVCNFSM6AAAAABC6WWOBSVHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM4DKMBWGY3TK>
>> .
>> You are receiving this because you authored the thread.Message ID:
>> ***@***.***>
>>
>
>
> --
> *Scott Wolf, DMA*
> Oxnard College - Professor of Music
> www.scottwolfguitar.com
>
> All Strings Considered <http://www.scottwolfguitar.com/podcast.html>
> Follow on Twitter - @allstrings <https://twitter.com/allstrings>
> Subscribe free on iTunes
> <https://itunes.apple.com/us/podcast/all-strings-considered/id562425515> -
> Subscribe - Please Rate!
>
>
>
>

-- 
*Scott Wolf, DMA*
Oxnard College - Professor of Music
www.scottwolfguitar.com

All Strings Considered <http://www.scottwolfguitar.com/podcast.html>
Follow on Twitter - @allstrings <https://twitter.com/allstrings>
Subscribe free on iTunes
<https://itunes.apple.com/us/podcast/all-strings-considered/id562425515> -
Subscribe - Please Rate!
